File: tests/test_opensubtitles.py
import io
import json
import os
import stat
import urllib.error

import pytest

from jetson_player.subtitles import opensubtitles as osub


def test_moviehash_of_zero_file_is_its_size(tmp_path):
    f = tmp_path / "zero.bin"
    f.write_bytes(b"\0" * 131072)
    assert osub.moviehash(str(f)) == "0000000000020000"


def test_moviehash_sums_head_and_tail_words(tmp_path):
    f = tmp_path / "v.bin"
    data = bytearray(200_000)
    data[0:8] = (1).to_bytes(8, "little")                 # 앞 64KB 첫 단어
    data[-8:] = (0xFFFFFFFFFFFFFFFF).to_bytes(8, "little")  # 뒤 64KB 마지막 단어 (오버플로 확인)
    f.write_bytes(bytes(data))
    assert osub.moviehash(str(f)) == f"{(200_000 + 1 + 0xFFFFFFFFFFFFFFFF) & 0xFFFFFFFFFFFFFFFF:016x}"
    small = tmp_path / "s.bin"
    small.write_bytes(b"x" * 1000)
    with pytest.raises(osub.OpenSubtitlesError):
        osub.moviehash(str(small))


@pytest.mark.parametrize("name, expected", [
    ("The.Movie.Name.2019.1080p.WEB-DL.x265-GRP.mkv", "The Movie Name 2019"),
    ("[SubGroup] Show_Name - 05 (1080p).mkv", "Show Name - 05"),
    ("드라마 12화.mp4", "드라마 12화"),
])
def test_query_from_filename(name, expected):
    assert osub.query_from_filename(name) == expected


SEARCH_JSON = {"data": [
    {"attributes": {"language": "en", "download_count": 900, "moviehash_match": False, "release": "Other.Release",
                    "feature_details": {"title": "Movie", "year": 2019}, "files": [{"file_id": 11, "file_name": "a.srt"}]}},
    {"attributes": {"language": "ko", "download_count": 10, "moviehash_match": True, "release": "Exact.Release",
                    "feature_details": {"title": "Movie", "year": 2019}, "files": [{"file_id": 22, "file_name": "b.ass"}]}},
    {"attributes": {"language": "ko", "files": []}},
]}


class FakeResponse(io.BytesIO):
    def __enter__(self):
        return self

    def __exit__(self, *a):
        return False


def fake_opener(routes, log):
    def opener(req, timeout=None):
        log.append((req.get_method(), req.full_url, dict(req.header_items()), req.data))
        for prefix, payload in routes:
            if req.full_url.startswith(prefix):
                if isinstance(payload, Exception):
                    raise payload
                return FakeResponse(payload if isinstance(payload, bytes) else json.dumps(payload).encode())
        raise AssertionError(req.full_url)
    return opener


def test_search_sorts_hash_match_first_and_sends_key(tmp_path):
    video = tmp_path / "Movie.2019.1080p.mkv"
    video.write_bytes(b"\0" * 200_000)
    calls = []
    client = osub.OpenSubtitlesClient("KEY", opener=fake_opener([(osub.API_BASE + "/subtitles", SEARCH_JSON)], calls))
    results = client.search(str(video), languages=("ko", "en"))
    assert [r.file_id for r in results] == [22, 11]
    assert results[0].hash_match and results[0].title == "Movie (2019)"
    method, url, headers, _ = calls[0]
    assert method == "GET" and "moviehash=" in url and "languages=en%2Cko" in url and "query=Movie+2019" in url
    assert headers["Api-key"] == "KEY" and headers["User-agent"].startswith("JetsonVideoPlayer")


def test_download_logs_in_and_fetches_link(tmp_path):
    calls = []
    routes = [(osub.API_BASE + "/login", {"token": "TOK"}),
              (osub.API_BASE + "/download", {"link": "https://dl.example/x.srt", "remaining": 5}),
              ("https://dl.example/", b"1\n00:00:01,000 --> 00:00:02,000\nhi\n")]
    client = osub.OpenSubtitlesClient("KEY", username="me", password="pw", opener=fake_opener(routes, calls))
    assert client.download(22).startswith(b"1\n")
    assert [c[1].split("/")[-1] for c in calls] == ["login", "download", "x.srt"]
    assert calls[1][2]["Authorization"] == "Bearer TOK" and json.loads(calls[1][3]) == {"file_id": 22}


def test_http_errors_become_readable(tmp_path):
    err = urllib.error.HTTPError(osub.API_BASE + "/download", 406, "x", {}, io.BytesIO(b'{"message": "quota exceeded"}'))
    client = osub.OpenSubtitlesClient("KEY", opener=fake_opener([(osub.API_BASE, err)], []))
    with pytest.raises(osub.OpenSubtitlesError, match="406 quota exceeded"):
        client.download(1)


def test_credentials_file_is_private(tmp_path, monkeypatch):
    monkeypatch.delenv("JVP_OPENSUBTITLES_KEY", raising=False)
    path = str(tmp_path / "cfg" / "os.json")
    osub.save_credentials({"api_key": "K", "username": "", "password": "p"}, path)
    assert stat.S_IMODE(os.stat(path).st_mode) == 0o600
    assert osub.load_credentials(path) == {"api_key": "K", "password": "p"}
    monkeypatch.setenv("JVP_OPENSUBTITLES_KEY", "ENV")
    assert osub.load_credentials(path)["api_key"] == "ENV"


def test_save_path(tmp_path):
    assert osub.subtitle_save_path("/v/movie.mkv", "ko", "x.ASS") == "/v/movie.ko.ass"
    assert osub.subtitle_save_path("/v/movie.mkv", "pt-BR", "x.zip") == "/v/movie.pt-br.srt"

File: app/tests/test_opensubtitles.cpp
// OpenSubtitles: tests/test_opensubtitles.py 이식 (로컬 가짜 HTTP 서버, 네트워크 사용 안 함)
#include "OpenSubtitles.h"
#include "Paths.h"
#include "fake_http_server.h"
#include "media_test_util.h"

#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QTest>

#include <sys/stat.h>

using namespace jvp::opensubtitles;

namespace {
void writeFile(const QString &path, const QByteArray &data)
{
    QFile f(path);
    QVERIFY(f.open(QIODevice::WriteOnly));
    f.write(data);
}

const QByteArray kSearchJson = R"({"data": [
    {"attributes": {"language": "en", "download_count": 900, "moviehash_match": false, "release": "Other.Release",
                    "feature_details": {"title": "Movie", "year": 2019}, "files": [{"file_id": 11, "file_name": "a.srt"}]}},
    {"attributes": {"language": "ko", "download_count": 10, "moviehash_match": true, "release": "Exact.Release",
                    "feature_details": {"title": "Movie", "year": 2019}, "files": [{"file_id": 22, "file_name": "b.ass"}]}},
    {"attributes": {"language": "ko", "files": []}}
]})";
} // namespace

class TestOpenSubtitles : public QObject {
    Q_OBJECT
    QString m_home;

private slots:
    void initTestCase()
    {
        m_home = testutil::isolateHome();
        qunsetenv("JVP_OPENSUBTITLES_KEY");
    }

    void moviehashOfZeroFileIsItsSize()
    {
        writeFile(m_home + "/zero.bin", QByteArray(131072, '\0'));
        QCOMPARE(moviehash(m_home + "/zero.bin"), QString("0000000000020000"));
    }

    void moviehashSumsHeadAndTailWords()
    {
        QByteArray data(200000, '\0');
        data[0] = 1;                             // 앞 64KB 첫 단어 = 1
        for (int i = 1; i <= 8; ++i)
            data[data.size() - i] = char(0xFF);  // 뒤 64KB 마지막 단어 = 0xFFFFFFFFFFFFFFFF (오버플로 확인)
        writeFile(m_home + "/v.bin", data);
        const quint64 expected = quint64(200000) + 1 + 0xFFFFFFFFFFFFFFFFULL;
        QCOMPARE(moviehash(m_home + "/v.bin"), QStringLiteral("%1").arg(expected, 16, 16, QLatin1Char('0')));
        writeFile(m_home + "/s.bin", QByteArray(1000, 'x'));
        QString err;
        QVERIFY(moviehash(m_home + "/s.bin", &err).isEmpty());
        QVERIFY(!err.isEmpty());
    }

    void queryFromFilenameCases()
    {
        QCOMPARE(queryFromFilename("The.Movie.Name.2019.1080p.WEB-DL.x265-GRP.mkv"), QString("The Movie Name 2019"));
        QCOMPARE(queryFromFilename("[SubGroup] Show_Name - 05 (1080p).mkv"), QString("Show Name - 05"));
        QCOMPARE(queryFromFilename(QStringLiteral("드라마 12화.mp4")), QStringLiteral("드라마 12화"));
    }

    void searchSortsHashMatchFirstAndSendsKey()
    {
        const QString video = m_home + "/Movie.2019.1080p.mkv";
        writeFile(video, QByteArray(200000, '\0'));
        FakeHttpServer server([](const FakeRequest &r) {
            if (r.target.startsWith("/api/v1/subtitles"))
                return FakeResponse{200, kSearchJson};
            return FakeResponse{404, "{}"};
        });
        OpenSubtitlesClient client("KEY");
        client.setApiBase(server.baseUrl() + "/api/v1");
        QList<SubtitleResult> results;
        QString error;
        bool done = false;
        client.search(video, {"ko", "en"}, [&](const QList<SubtitleResult> &r, const QString &e) {
            results = r;
            error = e;
            done = true;
        });
        QVERIFY(testutil::waitUntil([&] { return done; }));
        QVERIFY2(error.isEmpty(), qPrintable(error));
        QCOMPARE(results.size(), 2);
        QCOMPARE(results[0].fileId, 22);
        QCOMPARE(results[1].fileId, 11);
        QVERIFY(results[0].hashMatch);
        QCOMPARE(results[0].title, QString("Movie (2019)"));
        QCOMPARE(results[0].release, QString("Exact.Release"));
        QCOMPARE(results[0].fileName, QString("b.ass"));
        const FakeRequest &r = server.requests.first();
        QCOMPARE(r.method, QByteArray("GET"));
        QVERIFY2(r.target.contains("moviehash=") && r.target.contains("languages=en%2Cko") && r.target.contains("query=Movie+2019"),
                 r.target.constData());
        QVERIFY(r.target.indexOf("languages=") < r.target.indexOf("moviehash=")
                && r.target.indexOf("moviehash=") < r.target.indexOf("query="));   // 키 이름 순
        QCOMPARE(r.headers.value("api-key"), QByteArray("KEY"));
        QVERIFY(r.headers.value("user-agent").startsWith("JetsonVideoPlayer"));
        QCOMPARE(r.headers.value("accept"), QByteArray("application/json"));
    }

    void downloadLogsInAndFetchesLink()
    {
        FakeHttpServer *srv = nullptr;
        FakeHttpServer server([&](const FakeRequest &r) {
            if (r.target.endsWith("/login"))
                return FakeResponse{200, R"({"token": "TOK"})"};
            if (r.target.endsWith("/download"))
                return FakeResponse{200, QJsonDocument(QJsonObject{{"link", QString(srv->baseUrl() + "/dl/x.srt")}, {"remaining", 5}})
                                             .toJson(QJsonDocument::Compact)};
            if (r.target == "/dl/x.srt")
                return FakeResponse{200, "1\n00:00:01,000 --> 00:00:02,000\nhi\n", "text/plain"};
            return FakeResponse{404, "{}"};
        });
        srv = &server;
        OpenSubtitlesClient client("KEY", "me", "pw");
        client.setApiBase(server.baseUrl() + "/api/v1");
        QByteArray content;
        QString error;
        bool done = false;
        client.download(22, [&](const QByteArray &c, const QString &e) {
            content = c;
            error = e;
            done = true;
        });
        QVERIFY(testutil::waitUntil([&] { return done; }));
        QVERIFY2(error.isEmpty(), qPrintable(error));
        QVERIFY(content.startsWith("1\n"));
        QStringList last;
        for (const auto &r : server.requests)
            last << QString::fromUtf8(r.target.split('/').last());
        QCOMPARE(last, (QStringList{"login", "download", "x.srt"}));
        QCOMPARE(QJsonDocument::fromJson(server.requests[0].body).object(),
                 (QJsonObject{{"username", "me"}, {"password", "pw"}}));
        QCOMPARE(server.requests[1].headers.value("authorization"), QByteArray("Bearer TOK"));
        QCOMPARE(QJsonDocument::fromJson(server.requests[1].body).object(), (QJsonObject{{"file_id", 22}}));
        QCOMPARE(server.requests[1].headers.value("content-type"), QByteArray("application/json"));
        QVERIFY(!server.requests[2].headers.contains("api-key"));   // 임시 링크에는 키를 보내지 않음
        QCOMPARE(client.token(), QString("TOK"));
    }

    void httpErrorsBecomeReadable()
    {
        FakeHttpServer server([](const FakeRequest &) { return FakeResponse{406, R"({"message": "quota exceeded"})"}; });
        OpenSubtitlesClient client("KEY");
        client.setApiBase(server.baseUrl() + "/api/v1");
        QString error;
        bool done = false;
        client.download(1, [&](const QByteArray &, const QString &e) {
            error = e;
            done = true;
        });
        QVERIFY(testutil::waitUntil([&] { return done; }));
        QCOMPARE(error, QString("HTTP 406 quota exceeded"));

        // 연결 실패
        quint16 port;
        {
            QTcpServer tmp;
            tmp.listen(QHostAddress::LocalHost, 0);
            port = tmp.serverPort();
        }
        client.setApiBase(QStringLiteral("http://127.0.0.1:%1/api/v1").arg(port));
        done = false;
        client.download(1, [&](const QByteArray &, const QString &e) {
            error = e;
            done = true;
        });
        QVERIFY(testutil::waitUntil([&] { return done; }));
        QVERIFY2(error.startsWith(QStringLiteral("연결 실패")), qPrintable(error));

        OpenSubtitlesClient noKey("");
        done = false;
        noKey.download(1, [&](const QByteArray &, const QString &e) {
            error = e;
            done = true;
        });
        QVERIFY(testutil::waitUntil([&] { return done; }));
        QCOMPARE(error, QStringLiteral("API 키가 없습니다"));
    }

    void credentialsFileIsPrivate()
    {
        QVERIFY(credentialsFile().startsWith(m_home));   // 실제 ~/.config를 건드리지 않음
        const QString path = m_home + "/cfg/os.json";
        QVERIFY(saveCredentials({{"api_key", "K"}, {"username", ""}, {"password", "p"}}, path));
        struct stat st {};
        QCOMPARE(::stat(QFile::encodeName(path).constData(), &st), 0);
        QCOMPARE(int(st.st_mode & 0777), 0600);
        QCOMPARE(loadCredentials(path), (QMap<QString, QString>{{"api_key", "K"}, {"password", "p"}}));
        qputenv("JVP_OPENSUBTITLES_KEY", "ENV");
        QCOMPARE(loadCredentials(path).value("api_key"), QString("ENV"));
        qunsetenv("JVP_OPENSUBTITLES_KEY");
    }

    void savePath()
    {
        QCOMPARE(subtitleSavePath("/v/movie.mkv", "ko", "x.ASS"), QString("/v/movie.ko.ass"));
        QCOMPARE(subtitleSavePath("/v/movie.mkv", "pt-BR", "x.zip"), QString("/v/movie.pt-br.srt"));
        QCOMPARE(subtitleSavePath("/v/a.b/movie", "", ""), QString("/v/a.b/movie.sub.srt"));
    }

    void urlEncodeMatchesQuotePlus()
    {
        QCOMPARE(urlEncode({{"query", "Movie 2019"}, {"languages", "en,ko"}, {"x", "a+b/c~"}}),
                 QByteArray("query=Movie+2019&languages=en%2Cko&x=a%2Bb%2Fc~"));
    }
};

QTEST_GUILESS_MAIN(TestOpenSubtitles)
#include "test_opensubtitles.moc"
